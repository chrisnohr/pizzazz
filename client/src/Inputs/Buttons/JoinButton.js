import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

const JoinButton = ({ countdownValue, canJoin, joinGame }) => {
    if (countdownValue > 20 && canJoin) {
      return (
        <Button
          value="Join the current game"
          clickHandler={joinGame}
        />
      );
    }
    return null;
  };
  
  JoinButton.propTypes = {
    countdownValue: PropTypes.number.isRequired,
    canJoin: PropTypes.bool.isRequired,
    joinGame: PropTypes.func.isRequired,
  };
  
  export default JoinButton;