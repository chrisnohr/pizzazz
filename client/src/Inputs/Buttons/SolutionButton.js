import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';
import { getWordScore } from '../../gameFlow/solutions/findWinner';
import { letterValues } from '../../Constants';

const SolutionButton = ({ isValidIndex, scrabbleBoard, sendSolution }) => {
    if (isValidIndex) {
      const solution = scrabbleBoard.substring(0, isValidIndex);
      const potentialScore = getWordScore(solution, letterValues);
      const text = `Play ${solution.toUpperCase()} for ${potentialScore} points`;
      const clickHandler = () => sendSolution(solution, potentialScore, isValidIndex);
      return (
        <Button
          value={text}
          clickHandler={clickHandler}
        />
      );
    }
    return null;
  };
  
  SolutionButton.propTypes = {
    isValidIndex: PropTypes.number.isRequired,
    scrabbleBoard: PropTypes.string.isRequired,
    sendSolution: PropTypes.func.isRequired,
  };
  
  export default SolutionButton;