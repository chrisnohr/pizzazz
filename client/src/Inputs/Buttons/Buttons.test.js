import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Button from './Button';
import JoinButton from './JoinButton';
import SolutionButton from './SolutionButton';

describe('<Button />', () => {
  it('calls the clickHandler function on click', () => {
    const mockCallback = jest.fn();
    const wrapper = shallow((<Button clickHandler={mockCallback} value="Go" />));
    wrapper.dive().find('StyledComponent').simulate('click');
    expect(mockCallback).toHaveBeenCalledTimes(1);
  });
  it('matches the snapshot', () => {
    const mockCallback = jest.fn();
    const wrapper = shallow(<Button clickHandler={mockCallback} value="Go" />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('<JoinButton>', () => {
  it('renders a <Button /> if canJoin is true and countdownValue > 20', () => {
    const mockCallback = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin
        joinGame={mockCallback}
      />
    ));
    expect(wrapper.find(Button)).toHaveLength(1);
  });

  it('renders a button with the right value when canJoin is true and countdownValue > 20', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin
        joinGame={mockCallback}
      />
    ));
    expect(wrapper.dive().props().value).toEqual('Join the current game');
  });

  it('renders nothing if canJoin is false', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin={false}
        joinGame={mockCallback}
      />
    ));
    expect(wrapper.type()).toEqual(null);
  });

  it('renders nothing if countdownValue <= 20', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <JoinButton
        countdownValue={20}
        canJoin
        joinGame={mockCallback}
      />
    ));
    expect(wrapper.type()).toEqual(null);
  });

  it('matches the snapshot', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin
        joinGame={mockCallback}
      />
    ));
    expect(wrapper).toMatchSnapshot();
  });
});

describe('<SolutionButton />', () => {
  it('renders a <Button /> with the valid word and its score', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <SolutionButton
        isValidIndex={7}
        scrabbleBoard="pizzazz"
        sendSolution={mockCallback}
      />
    ));
    const wrapper2 = shallow((
      <SolutionButton
        isValidIndex={5}
        scrabbleBoard="pizza00"
        sendSolution={mockCallback}
      />
    ));
    expect(wrapper.dive().props().value).toEqual('Play PIZZAZZ for 46 points');
    expect(wrapper2.dive().props().value).toEqual('Play PIZZA for 26 points');
  });

  it('does not render a <Button /> when isValidIndex is zero', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <SolutionButton
        isValidIndex={0}
        scrabbleBoard="pizarro"
        sendSolution={mockCallback}
      />
    ));
    expect(wrapper.find(Button).exists()).toEqual(false);
  });

  it('matches the snapshot', () => {
    const mockCallback = el => el;
    const wrapper = shallow((
      <SolutionButton
        isValidIndex={7}
        scrabbleBoard="pizzazz"
        sendSolution={mockCallback}
      />
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
