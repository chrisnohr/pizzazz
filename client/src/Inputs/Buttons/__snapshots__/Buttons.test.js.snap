// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    clickHandler={[MockFunction]}
    value="Go"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "autoFocus": true,
      "onClick": [MockFunction],
      "readOnly": true,
      "type": "button",
      "value": "Go",
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bwzfXH",
        "isStatic": true,
        "rules": Array [
          "
  display: inline-block;
  vertical-align: middle;
  font-size: ",
          "4vw",
          ";
  font-weight: ",
          "300",
          ";
  padding: 3vw 6vw;
  margin-bottom: ",
          "9vw",
          ";
  color: ",
          "#7A828A",
          ";
  border: .5px solid ",
          "#CED4DA",
          ";
  background-color: ",
          "#FFF",
          ";
  box-shadow: ",
          "1.8px 2.4px 1.8px rgba(0, 0, 0, 0.05)",
          ";
  @media screen and (min-width: 445px) {
    margin-bottom: ",
          "3rem",
          ";
    font-size: 1.1rem;
    padding: ",
          "0.8rem 1.6rem",
          ";
  };
  :focus {
    outline: 0;
    border: 1px solid ",
          "#BDEAFE",
          ";
  }
",
          "
  border-radius: .3rem;
  text-align: center;
  :active {
    box-shadow: inset ",
          "1.8px 2.4px 1.8px rgba(0, 0, 0, 0.05)",
          ";
  }
",
        ],
      },
      "displayName": "Styled(styled.input)",
      "foldedComponentIds": Array [
        "sc-bdVaJa",
      ],
      "render": [Function],
      "styledComponentId": "sc-bwzfXH",
      "target": "input",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "autoFocus": true,
        "onClick": [MockFunction],
        "readOnly": true,
        "type": "button",
        "value": "Go",
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": true,
          "rules": Array [
            "
  display: inline-block;
  vertical-align: middle;
  font-size: ",
            "4vw",
            ";
  font-weight: ",
            "300",
            ";
  padding: 3vw 6vw;
  margin-bottom: ",
            "9vw",
            ";
  color: ",
            "#7A828A",
            ";
  border: .5px solid ",
            "#CED4DA",
            ";
  background-color: ",
            "#FFF",
            ";
  box-shadow: ",
            "1.8px 2.4px 1.8px rgba(0, 0, 0, 0.05)",
            ";
  @media screen and (min-width: 445px) {
    margin-bottom: ",
            "3rem",
            ";
    font-size: 1.1rem;
    padding: ",
            "0.8rem 1.6rem",
            ";
  };
  :focus {
    outline: 0;
    border: 1px solid ",
            "#BDEAFE",
            ";
  }
",
            "
  border-radius: .3rem;
  text-align: center;
  :active {
    box-shadow: inset ",
            "1.8px 2.4px 1.8px rgba(0, 0, 0, 0.05)",
            ";
  }
",
          ],
        },
        "displayName": "Styled(styled.input)",
        "foldedComponentIds": Array [
          "sc-bdVaJa",
        ],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": "input",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<JoinButton> matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <JoinButton
    canJoin={true}
    countdownValue={21}
    joinGame={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "clickHandler": [Function],
      "value": "Join the current game",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "clickHandler": [Function],
        "value": "Join the current game",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<SolutionButton /> matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SolutionButton
    isValidIndex={7}
    scrabbleBoard="pizzazz"
    sendSolution={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "clickHandler": [Function],
      "value": "Play PIZZAZZ for 46 points",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "clickHandler": [Function],
        "value": "Play PIZZAZZ for 46 points",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
